### PR TITLE
8194823: Serial GC does not account GCs caused by TLAB allocation in GC overhead limit

### PR DESCRIPTION
**Pull request by lgxbslgx**
Hi all,

This patch enables the gc overhead limit when allocating TLAB in serial gc.
The main modification is at `GenCollectedHeap::allocate_new_tlab` and the other 
files only adjust the parameters of the method `allocate_new_tlab`.

Thanks for the review.

Best Regards,
-- Guoxiong

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8194823](https://bugs.openjdk.org/browse/JDK-8194823): Serial GC does not account GCs caused by TLAB allocation in GC overhead limit

### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to [f19d3213](https://git.openjdk.org/jdk/pull/14120/files/f19d32132467c13727f0a1082fa1885601e12461)
 * [Ivan Walulya](https://openjdk.org/census#iwalulya) (@walulyai - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14120/head:pull/14120` \
`$ git checkout pull/14120`

Update a local copy of the PR: \
`$ git checkout pull/14120` \
`$ git pull https://git.openjdk.org/jdk.git pull/14120/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14120`

View PR using the GUI difftool: \
`$ git pr show -t 14120`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14120.diff">https://git.openjdk.org/jdk/pull/14120.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14120#issuecomment-1560912818)

### CODE REVIEW

File: src/hotspot/share/gc/parallel/parallelScavengeHeap.cpp [hunk 1]
```diff
@@ -514,7 +514,8 @@ size_t ParallelScavengeHeap::unsafe_max_tlab_alloc(Thread* thr) const {
   return young_gen()->eden_space()->unsafe_max_tlab_alloc(thr);
 }
 
-HeapWord* ParallelScavengeHeap::allocate_new_tlab(size_t min_size, size_t requested_size, size_t* actual_size) {
+HeapWord* ParallelScavengeHeap::allocate_new_tlab(size_t min_size, size_t requested_size, size_t* actual_size,
+                                                  bool* gc_overhead_limit_was_exceeded) {
```
Comment (tschatzl) [inline src/hotspot/share/gc/parallel/parallelScavengeHeap.cpp:518]:
```suggestion
HeapWord* ParallelScavengeHeap::allocate_new_tlab(size_t min_size,
                                                  size_t requested_size,
                                                  size_t* actual_size,
                                                  bool* gc_overhead_limit_was_exceeded) {
```
Parameter list formatting

File: src/hotspot/share/gc/parallel/parallelScavengeHeap.hpp [hunk 2]
```diff
@@ -102,7 +102,8 @@ class ParallelScavengeHeap : public CollectedHeap {
   HeapWord* allocate_old_gen_and_record(size_t word_size);
 
  protected:
-  HeapWord* allocate_new_tlab(size_t min_size, size_t requested_size, size_t* actual_size) override;
+  HeapWord* allocate_new_tlab(size_t min_size, size_t requested_size, size_t* actual_size,
+                              bool* gc_overhead_limit_was_exceeded) override;
```
Comment (tschatzl) [inline src/hotspot/share/gc/parallel/parallelScavengeHeap.hpp:106]:
```suggestion
  HeapWord* allocate_new_tlab(size_t min_size,
                              size_t requested_size,
                              size_t* actual_size,
                              bool* gc_overhead_limit_was_exceeded) override;
```
Parameter formatting

File: src/hotspot/share/gc/shared/memAllocator.cpp [hunk 3]
```diff
@@ -321,7 +321,8 @@ HeapWord* MemAllocator::mem_allocate_inside_tlab_slow(Allocation& allocation) co
   // Allocate a new TLAB requesting new_tlab_size. Any size
   // between minimal and new_tlab_size is accepted.
   size_t min_tlab_size = ThreadLocalAllocBuffer::compute_min_size(_word_size);
-  mem = Universe::heap()->allocate_new_tlab(min_tlab_size, new_tlab_size, &allocation._allocated_tlab_size);
+  mem = Universe::heap()->allocate_new_tlab(min_tlab_size, new_tlab_size, &allocation._allocated_tlab_size,
+                                            &allocation._overhead_limit_exceeded);
```
Comment (tschatzl) [inline src/hotspot/share/gc/shared/memAllocator.cpp:325]:
```suggestion
  mem = Universe::heap()->allocate_new_tlab(min_tlab_size,
                                            new_tlab_size,
                                            &allocation._allocated_tlab_size,
                                            &allocation._overhead_limit_exceeded);
```
Parameter formatting

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp [hunk 4]
```diff
@@ -531,7 +531,8 @@ class ShenandoahHeap : public CollectedHeap {
 
   void notify_mutator_alloc_words(size_t words, bool waste);
 
-  HeapWord* allocate_new_tlab(size_t min_size, size_t requested_size, size_t* actual_size) override;
+  HeapWord* allocate_new_tlab(size_t min_size, size_t requested_size, size_t* actual_size,
+                              bool* gc_overhead_limit_was_exceeded) override;
```
Comment (tschatzl) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp:535]:
```suggestion
  HeapWord* allocate_new_tlab(size_t min_size,
                              size_t requested_size,
                              size_t* actual_size,
                              bool* gc_overhead_limit_was_exceeded) override;
```

File: src/hotspot/share/gc/x/xCollectedHeap.cpp [hunk 5]
```diff
@@ -146,7 +146,8 @@ bool XCollectedHeap::requires_barriers(stackChunkOop obj) const {
   return false;
 }
 
-HeapWord* XCollectedHeap::allocate_new_tlab(size_t min_size, size_t requested_size, size_t* actual_size) {
+HeapWord* XCollectedHeap::allocate_new_tlab(size_t min_size, size_t requested_size, size_t* actual_size,
+                                            bool* gc_overhead_limit_was_exceeded) {
```
Comment (tschatzl) [inline src/hotspot/share/gc/x/xCollectedHeap.cpp:150]:
```suggestion
HeapWord* XCollectedHeap::allocate_new_tlab(size_t min_size,
                                            size_t requested_size,
                                            size_t* actual_size,
                                            bool* gc_overhead_limit_was_exceeded) {
```

File: src/hotspot/share/gc/z/zCollectedHeap.cpp [hunk 6]
```diff
@@ -141,7 +141,8 @@ bool ZCollectedHeap::requires_barriers(stackChunkOop obj) const {
   return ZContinuation::requires_barriers(&_heap, obj);
 }
 
-HeapWord* ZCollectedHeap::allocate_new_tlab(size_t min_size, size_t requested_size, size_t* actual_size) {
+HeapWord* ZCollectedHeap::allocate_new_tlab(size_t min_size, size_t requested_size, size_t* actual_size,
+                                            bool* gc_overhead_limit_was_exceeded) {
```
Comment (tschatzl) [inline src/hotspot/share/gc/z/zCollectedHeap.cpp:145]:
```suggestion
HeapWord* ZCollectedHeap::allocate_new_tlab(size_t min_size,
                                            size_t requested_size,
                                            size_t* actual_size,
                                            bool* gc_overhead_limit_was_exceeded) {
```

**Review by tschatzl — APPROVED** [review]
Lgtm apart from parameter list formatting issues. Please fix before committing.

It's a bit unfortunate that support for this feature for one collector causes so many changes everywhere else. (But also indicates an issue with all other collectors not supporting it ;)). Maybe the parameters could be wrapped into something like an "AllocationRequest", but this is a) a separate issue, and b) needs to be discussed first with others.

File: src/hotspot/share/gc/parallel/parallelScavengeHeap.cpp (hunk 1 above)
Comment (lgxbslgx) [reply src/hotspot/share/gc/parallel/parallelScavengeHeap.cpp]: Fixed.

File: src/hotspot/share/gc/parallel/parallelScavengeHeap.hpp (hunk 2 above)
Comment (lgxbslgx) [reply src/hotspot/share/gc/parallel/parallelScavengeHeap.hpp]: Fixed.

File: src/hotspot/share/gc/shared/memAllocator.cpp (hunk 3 above)
Comment (lgxbslgx) [reply src/hotspot/share/gc/shared/memAllocator.cpp]: Fixed.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp (hunk 4 above)
Comment (lgxbslgx) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp]: Fixed.

File: src/hotspot/share/gc/x/xCollectedHeap.cpp (hunk 5 above)
Comment (lgxbslgx) [reply src/hotspot/share/gc/x/xCollectedHeap.cpp]: Fixed.

File: src/hotspot/share/gc/z/zCollectedHeap.cpp (hunk 6 above)
Comment (lgxbslgx) [reply src/hotspot/share/gc/z/zCollectedHeap.cpp]: Fixed.

**Comment by lgxbslgx** [comment]
@tschatzl Thanks for the review. I fixed the formatting issues just now.

**Review by walulyai — APPROVED** [review]

**Comment by albertnetymk** [comment]
Do you have a small example to trigger `java.lang.OutOfMemoryError: GC Overhead Limit Exceeded` for Serial GC. I was under the impression that Serial doesn't support `UseGCOverheadLimit`.

**Comment by albertnetymk** [comment]
> The UseGCOverheadLimit is only used when UseAdaptiveSizePolicy is true. Is it intentional?

I can't see any dependency btw them after skimming through their specification. Worth investigation in its own ticket/PR.

Some background which might be useful, "UseGCOverheadLimit can work independently of UseAdaptiveSizePolicy" from https://bugs.openjdk.org/browse/JDK-8212206

> should I remove the UseGCOverheadLimit related code in serial GC?

I am leaned towards removing it, as it is effectively dead code.

**Comment by lgxbslgx** [comment]
> I can't see any dependency btw them after skimming through their specification. Worth investigation in its own ticket/PR.
>
> Some background which might be useful, "UseGCOverheadLimit can work independently of UseAdaptiveSizePolicy" from https://bugs.openjdk.org/browse/JDK-8212206

Filed https://bugs.openjdk.org/browse/JDK-8308983 to follow up.

> I am leaned towards removing it, as it is effectively dead code.

I also think it is good to remove it.
If hearing no objection in the next several days, I will submit another issue&PR to remove it and close this one.
Or I should remove it in this PR? Need to be confirmed here.

**Comment by albertnetymk** [comment]
I think it's best to do the removing in another ticket/PR and link the two tickets.

**Pull request closed without merging**